### PR TITLE
WIP: Make the scenarios load without complaints

### PR DESCRIPTION
**Pull request by MartinGuehmann**
Fix the text files of the scenarios so that they load without complaints.

However the Alex scenario still crashes on load.

### CODE REVIEW

**Comment by LynxAbraxas** [comment]
Thanks for the fixes, tested these but still get 
```
 civilisation.txt:1500: CANADIANS_CITY_1 not in string database. Token: 257, TokenName: 257, Next: 565

civilisation.txt:1530: Missing open brace
``` 
or similar, which I can ignore to get to the launch button but at the end of loading the game when the progress bar is full and vanishes the game still crashes for nuke, samurai and alex, AE mod works though.

**Comment by LynxAbraxas** [comment]
@madand How does it work for your setup?

**Comment by MartinGuehmann** [comment]
If I remember correctly your CD is German, so could you go to /ctp2_code/ctp/civpath.txt and change the line english to german, and then try again?

As it looks like we only have all the German files in the repository. And meaybe the Chinese ones, too.

**Comment by madand** [comment]
Samurai: works OK, played for several turns.
Alexander, Nuclear: crash when map is almost loaded and bottom UI pane is already shown.
```
[1]    22229 illegal hardware instruction  ./ctp2 -nointromovie
```

**Comment by MartinGuehmann** [comment]
@madand what's your language version?

**Comment by madand** [comment]
English.

**Comment by LynxAbraxas** [comment]
Many thanks guys for your input. Got the samurai working with the English files. Jesus, that is catchy. Played it for more than a 100 turns... On my 4th attempt to defeat the monster blocking the way to the Star Road I got `Messagebox(Slic Error): In object SSMoveStuff_F, function _TerrainType: Wrong kind of builtin variable` when I lost Akimbo but still had the magician, which I moved off and was able to continue the game but had no chance any more.

**Comment by LynxAbraxas** [comment]
> If I remember correctly your CD is German, so could you go to /ctp2_code/ctp/civpath.txt and change the line english to german, and then try again?

Tested, but ended up with a different CPT2 start screen
ss_2020-04-02_00:46:42
so guess the DI is not properly set up for using the German language files, though they come from the German CD version.

**Comment by LynxAbraxas** [comment]
> Samurai: works OK, played for several turns.
Alexander, Nuclear: crash when map is almost loaded and bottom UI pane is already shown.

Yes, same for me.

**Comment by LynxAbraxas** [comment]
Just noticed that the scenario descriptions are in German for me
ss_2020-04-02_00:54:12
even if english is set in `civpath.txt` and everything else as well as sounds are in English.

NB: I guess the boxes at the end of strings are the `\r` which are still in the txt-files.

**Comment by MartinGuehmann** [comment]
> > If I remember correctly your CD is German, so could you go to /ctp2_code/ctp/civpath.txt and change the line english to german, and then try again?
> 
> Tested, but ended up with a different CPT2 start screen
> ss_2020-04-02_00:46:42
> so guess the DI is not properly set up for using the German language files, though they come from the German CD version.

I guess this will go away when #288 has been merged. I guess it uses the German ldl files from the CD.

**Comment by MartinGuehmann** [comment]
> Just noticed that the scenario descriptions are in German for me
> ss_2020-04-02_00:54:12
> even if english is set in `civpath.txt` and everything else as well as sounds are in English.
> 
> NB: I guess the boxes at the end of strings are the `\r` which are still in the txt-files.

The text come from the packlist.txt file in the root scenario folder. Of course this is not very nice since these are not localized. That means they are now for everybody in German.

I guess the content of the packlist.txt file is printed here. While the other text files contain quoted strings. So with those the end of lines don't matter.

**Comment by LynxAbraxas** [comment]
> I guess this will go away when #288 has been merged. I guess it uses the German ldl files from the CD.

Yes, it does, many thanks for the pointer.

**Comment by LynxAbraxas** [comment]
Still, nuke and alex crash close to map load.

**Comment by MartinGuehmann** [comment]
Well, Alex also crashes for me, but I can load Nuke, with the debug version, actually I get some trade related asserts.

**Comment by LynxAbraxas** [comment]
> actually I get some trade related asserts.

I'll check.

**Comment by LynxAbraxas** [comment]
Hm, for me it asserts as:
```
Assertion (start < m_code + m_codeSize) Failed in File:SlicSegment.cpp, Line:649
Messagebox(Assert): Assertion (start < m_code + m_codeSize) Failed in File:SlicSegment.cpp, Line:649
```

**Comment by MartinGuehmann** [comment]
> Hm, for me it asserts as:
> 
> ```
> Assertion (start < m_code + m_codeSize) Failed in File:SlicSegment.cpp, Line:649
> Messagebox(Assert): Assertion (start < m_code + m_codeSize) Failed in File:SlicSegment.cpp, Line:649
> ```

That could be another save game incompatibility.

**Comment by LynxAbraxas** [comment]
> > Hm, for me it asserts as:
> > ```
> > Assertion (start < m_code + m_codeSize) Failed in File:SlicSegment.cpp, Line:649
> > Messagebox(Assert): Assertion (start < m_code + m_codeSize) Failed in File:SlicSegment.cpp, Line:649
> > ```
> 
> That could be another 64 bit save game incompatibility.

Odd though, that it does not assert for you at that point. What is the assertion that you get?

**Comment by MartinGuehmann** [comment]
Just keep in mind that you are trying to load a savegame with a 64 bit executable, while the Windows executable is 32 bits. And there might still be something in the save game format I haven't caught, simply because there is too much code to check.

Sorry, right now I can only give you the one for the Alex scenario. The debug version does not run on wine:

```
  civ3_main.cpp@1262: Exception: 'Integer Divide By Zero' thrown.
  civ3_main.cpp@1264: Exception Stack Trace:
  0x00a0d1ad  [public: void __thiscall Governor::ComputeDesiredUnits(void) + 0xc4d]
  0x00a10d94  [public: void __thiscall Governor::FillEmptyBuildQueues(bool) + 0x84]
  0x00a30fe3  [public: static void __cdecl CtpAi::FinishBeginTurn(int) + 0x153]
  0x00d07725  [private: virtual enum GAME_EVENT_HOOK_DISPOSITION __thiscall AIFinishBeginTurnEvent::GEVHookCallback(enum GAME_EVENT,class GameEventArgList *) + 0x45]
  0x00bdb3ca  [public: enum GAME_EVENT_ERR __thiscall GameEventHook::Activate(class GameEventArgList *,int,int &)const  + 0xfa]
  0x00bde826  [public: enum GAME_EVENT_ERR __thiscall GameEventManager::ActivateHook(enum GAME_EVENT,class GameEventArgList *,int,int &) + 0x86]
  0x00bd6145  [public: enum GAME_EVENT_ERR __thiscall GameEvent::Process(void) + 0x65]
  0x00be18f6  [public: enum GAME_EVENT_ERR __thiscall GameEventManager::ProcessHead(void) + 0x76]
  0x00be180e  [public: enum GAME_EVENT_ERR __thiscall GameEventManager::Process(void) + 0x8e]
  0x00be0930  [public: enum GAME_EVENT_ERR __thiscall GameEventManager::ArglistAddEvent(enum GAME_EVENT_INSERT,enum GAME_EVENT,class GameEventArgList *) + 0x210]
  0x00bdebb1  [private: enum GAME_EVENT_ERR __thiscall GameEventManager::AddEvent(enum GAME_EVENT_INSERT,enum GAME_EVENT,enum GAME_EVENT_ARGUMENT const *,void const * *) + 0x1c1]
  0x00bdf5b3  [public: enum GAME_EVENT_ERR __thiscall GameEventManager::AddEvent(enum GAME_EVENT_INSERT,enum GAME_EVENT,enum GAME_EVENT_ARGUMENT,int,enum GAME_EVENT_ARGUMENT,int,enum GAME_EVENT_ARGUMENT) + 0x63]
  0x01476080  [public: int __thiscall CivApp::InitializeGame(class CivArchive &) + 0x770]
  0x00bf2c53  [public: unsigned int __thiscall GameFile::Restore(char const *) + 0xd63]
  0x00bf37f4  [public: static void __cdecl GameFile::RestoreGame(char const *) + 0x34]
  0x01478639  [public: int __thiscall CivApp::LoadSavedGame(char const *) + 0x159]
  0x014715a5  [public: virtual void __thiscall LoadSaveGameAction::Execute(class aui_Control *,unsigned int,unsigned int) + 0x35]
  0x011753c1  [public: void __thiscall aui_UI::HandleActions(void) + 0x81]
  0x01178596  [public: virtual enum AUI_ERRCODE __thiscall aui_UI::Process(void) + 0x86]
  0x0147a27a  [private: int __thiscall CivApp::ProcessUI(unsigned int,unsigned int &) + 0x27a]
  0x0147973f  [public: int __thiscall CivApp::Process(void) + 0x10f]
  0x0146653b  [int __stdcall CivMain(struct HINSTANCE__ *,struct HINSTANCE__ *,char *,int) + 0x41b]
  0x0146b3ae  [WinMain@16 + 0xae]
  0x014ba38e  [int __cdecl invoke_main(void) + 0x1e]
  0x014ba227  [int __cdecl __scrt_common_main_seh(void) + 0x157]
  0x014ba0bd  [int __cdecl __scrt_common_main(void) + 0xd]
  0x014ba408  [WinMainCRTStartup + 0x8]
  0x75816359  [$R002030 + 0x74032fc9]
  0x774e7b74  [$R002030 + 0x75d047e4]

```

That is quite clear what it is. Somebody decided that it is a cool idea to give some boats a transport capacity of exactly zero units. I still don't know whether these transporters are supposed to be transporters or not. Obviously, if not than also the build lists must be adapted.

**Comment by MartinGuehmann** [comment]
Here is the assert with some debug log before

```
    objpool.cpp@99  : ObjPool: Deleting object id 0xc0000000
   slicfunc.cpp@539 : Slic_Text: Your highness, the {player[1].civ_name_plural} have declared war on us!
    objpool.cpp@92  : ObjPool: Inserting object id 0xc0000000
   slicfunc.cpp@539 : Slic_Text: The {player[0].civ_name_plural} have gone to war with the {player[1].civ_name_plural}.
   slicfunc.cpp@539 : Slic_Text: {player[0].leader_name} has cancelled the trade route carrying {good[0].name} from {city[0].name} to {city[1].name}, which was worth {gold[0].value} food, {gold[1].value} production, {gold[2].value} gold.
   slicfunc.cpp@704 : Slic_EyePoint: (58,78,0)
   slicfunc.cpp@539 : Slic_Text: {player[0].leader_name} has cancelled the trade route carrying {good[0].name} from {city[0].name} to {city[1].name}, which was worth {gold[0].value} food, {gold[1].value} production, {gold[2].value} gold.
   slicfunc.cpp@704 : Slic_EyePoint: (58,78,0)
    objpool.cpp@92  : ObjPool: Inserting object id 0xc0000001
       unit.cpp@167 : Unit::RemoveAllReferences: id: 0x100000fc cause: 47
     player.cpp@1242: Player::RemoveUnitReference(100000fc, 47)
    objpool.cpp@105 : ObjPool: Deleting object id 0x100000fc
   slicfunc.cpp@539 : Slic_Text: {player[0].leader_name} has cancelled the trade route carrying {good[0].name} from {city[0].name} to {city[1].name}, which was worth {gold[0].value} food, {gold[1].value} production, {gold[2].value} gold.
   slicfunc.cpp@704 : Slic_EyePoint: (36,66,0)
   slicfunc.cpp@539 : Slic_Text: {player[0].leader_name} has cancelled the trade route carrying {good[0].name} from {city[0].name} to {city[1].name}, which was worth {gold[0].value} food, {gold[1].value} production, {gold[2].value} gold.
   slicfunc.cpp@704 : Slic_EyePoint: (36,66,0)
    objpool.cpp@92  : ObjPool: Inserting object id 0xc0000002
    c3debug.cpp@303 : Assertion (m_traderUnits->Num() > 0) Failed in File:c:\activision\civctp2\ctp2_code\gs\gameobj\player.cpp, Line:3246
    c3debug.cpp@304 : Stack Trace: '  0x0149f5ec  [char * __cdecl c3debug_StackTrace(void) + 0x3c]
  0x01497342  [void __cdecl c3debug_Assert(char const *,char const *,int) + 0x72]
  0x00cf597f  [public: void __thiscall Player::KillATrader(void) + 0x6f]
  0x00cf9761  [public: void __thiscall Player::RemoveTradeRoute(class TradeRoute,enum CAUSE_KILL_TRADE_ROUTE) + 0x91]
  0x00d23f0c  [public: void __thiscall TradeRoute::KillRoute(enum CAUSE_KILL_TRADE_ROUTE) + 0x3bc]
  0x00c7981f  [public: void __thiscall TradeRoute::Kill(enum CAUSE_KILL_TRADE_ROUTE) + 0x2f]
  0x00d1f6d9  [private: virtual enum GAME_EVENT_HOOK_DISPOSITION __thiscall KillTradeRouteEvent::GEVHookCallback(enum GAME_EVENT,class GameEventArgList *) + 0x69]
  0x00bdb3fa  [public: enum GAME_EVENT_ERR __thiscall GameEventHook::Activate(class GameEventArgList *,int,int &)const  + 0xfa]
  0x00bde856  [public: enum GAME_EVENT_ERR __thiscall GameEventManager::ActivateHook(enum GAME_EVENT,class GameEventArgList *,int,int &) + 0x86]
  0x00bd6175  [public: enum GAME_EVENT_ERR __thiscall GameEvent::Process(void) + 0x65]
  0x00be1926  [public: enum GAME_EVENT_ERR __thiscall GameEventManager::ProcessHead(void) + 0x76]
  0x00be183e  [public: enum GAME_EVENT_ERR __thiscall GameEventManager::Process(void) + 0x8e]
  0x014765b1  [public: int __thiscall CivApp::InitializeGame(class CivArchive &) + 0x6d1]
  0x00bf2cb3  [public: unsigned int __thiscall GameFile::Restore(char const *) + 0xd63]
  0x00bf3854  [public: static void __cdecl GameFile::RestoreGame(char const *) + 0x34]
  0x01478c09  [public: int __thiscall CivApp::LoadSavedGame(char const *) + 0x159]
  0x01471b75  [public: virtual void __thiscall LoadSaveGameAction::Execute(class aui_Control *,unsigned int,unsigned int) + 0x35]
  0x01175461  [public: void __thiscall aui_UI::HandleActions(void) + 0x81]
  0x01178636  [public: virtual enum AUI_ERRCODE __thiscall aui_UI::Process(void) + 0x86]
  0x0147a84a  [private: int __thiscall CivApp::ProcessUI(unsigned int,unsigned int &) + 0x27a]
  0x01479d0f  [public: int __thiscall CivApp::Process(void) + 0x10f]
  0x01466b0b  [int __stdcall CivMain(struct HINSTANCE__ *,struct HINSTANCE__ *,char *,int) + 0x41b]
  0x0146b97e  [WinMain@16 + 0xae]
  0x014ba97e  [int __cdecl invoke_main(void) + 0x1e]
  0x014ba817  [int __cdecl __scrt_common_main_seh(void) + 0x157]
  0x014ba6ad  [int __cdecl __scrt_common_main(void) + 0xd]
  0x014ba9f8  [WinMainCRTStartup + 0x8]
  0x74e06359  [$R002030 + 0x73622fc9]
  0x775a7b74  [$R002030 + 0x75dc47e4]
'
   slicfunc.cpp@539 : Slic_Text: {player[0].leader_name} has cancelled the trade route carrying {good[0].name} from {city[0].name} to {city[1].name}, which was worth {gold[0].value} food, {gold[1].value} production, {gold[2].value} gold.
   slicfunc.cpp@704 : Slic_EyePoint: (44,70,0)
   slicfunc.cpp@539 : Slic_Text: {player[0].leader_name} has cancelled the trade route carrying {good[0].name} from {city[0].name} to {city[1].name}, which was worth {gold[0].value} food, {gold[1].value} production, {gold[2].value} gold.
   slicfunc.cpp@704 : Slic_EyePoint: (44,70,0)
    objpool.cpp@92  : ObjPool: Inserting object id 0xc0000003
    c3debug.cpp@303 : Assertion (m_traderUnits->Num() > 0) Failed in File:c:\activision\civctp2\ctp2_code\gs\gameobj\player.cpp, Line:3246
    c3debug.cpp@304 : Stack Trace: '  0x0149f5ec  [char * __cdecl c3debug_StackTrace(void) + 0x3c]
  0x01497342  [void __cdecl c3debug_Assert(char const *,char const *,int) + 0x72]
  0x00cf597f  [public: void __thiscall Player::KillATrader(void) + 0x6f]
  0x00cf9761  [public: void __thiscall Player::RemoveTradeRoute(class TradeRoute,enum CAUSE_KILL_TRADE_ROUTE) + 0x91]
  0x00d23f0c  [public: void __thiscall TradeRoute::KillRoute(enum CAUSE_KILL_TRADE_ROUTE) + 0x3bc]
  0x00c7981f  [public: void __thiscall TradeRoute::Kill(enum CAUSE_KILL_TRADE_ROUTE) + 0x2f]
  0x00d1f6d9  [private: virtual enum GAME_EVENT_HOOK_DISPOSITION __thiscall KillTradeRouteEvent::GEVHookCallback(enum GAME_EVENT,class GameEventArgList *) + 0x69]
  0x00bdb3fa  [public: enum GAME_EVENT_ERR __thiscall GameEventHook::Activate(class GameEventArgList *,int,int &)const  + 0xfa]
  0x00bde856  [public: enum GAME_EVENT_ERR __thiscall GameEventManager::ActivateHook(enum GAME_EVENT,class GameEventArgList *,int,int &) + 0x86]
  0x00bd6175  [public: enum GAME_EVENT_ERR __thiscall GameEvent::Process(void) + 0x65]
  0x00be1926  [public: enum GAME_EVENT_ERR __thiscall GameEventManager::ProcessHead(void) + 0x76]
  0x00be183e  [public: enum GAME_EVENT_ERR __thiscall GameEventManager::Process(void) + 0x8e]
  0x014765b1  [public: int __thiscall CivApp::InitializeGame(class CivArchive &) + 0x6d1]
  0x00bf2cb3  [public: unsigned int __thiscall GameFile::Restore(char const *) + 0xd63]
  0x00bf3854  [public: static void __cdecl GameFile::RestoreGame(char const *) + 0x34]
  0x01478c09  [public: int __thiscall CivApp::LoadSavedGame(char const *) + 0x159]
  0x01471b75  [public: virtual void __thiscall LoadSaveGameAction::Execute(class aui_Control *,unsigned int,unsigned int) + 0x35]
  0x01175461  [public: void __thiscall aui_UI::HandleActions(void) + 0x81]
  0x01178636  [public: virtual enum AUI_ERRCODE __thiscall aui_UI::Process(void) + 0x86]
  0x0147a84a  [private: int __thiscall CivApp::ProcessUI(unsigned int,unsigned int &) + 0x27a]
  0x01479d0f  [public: int __thiscall CivApp::Process(void) + 0x10f]
  0x01466b0b  [int __stdcall CivMain(struct HINSTANCE__ *,struct HINSTANCE__ *,char *,int) + 0x41b]
  0x0146b97e  [WinMain@16 + 0xae]
  0x014ba97e  [int __cdecl invoke_main(void) + 0x1e]
  0x014ba817  [int __cdecl __scrt_common_main_seh(void) + 0x157]
  0x014ba6ad  [int __cdecl __scrt_common_main(void) + 0xd]
  0x014ba9f8  [WinMainCRTStartup + 0x8]
  0x74e06359  [$R002030 + 0x73622fc9]
  0x775a7b74  [$R002030 + 0x75dc47e4]
'
   slicfunc.cpp@539 : Slic_Text: {player[0].leader_name} has cancelled the trade route carrying {good[0].name} from {city[0].name} to {city[1].name}, which was worth {gold[0].value} food, {gold[1].value} production, {gold[2].value} gold.
   slicfunc.cpp@704 : Slic_EyePoint: (45,100,0)
   slicfunc.cpp@539 : Slic_Text: {player[0].leader_name} has cancelled the trade route carrying {good[0].name} from {city[0].name} to {city[1].name}, which was worth {gold[0].value} food, {gold[1].value} production, {gold[2].value} gold.
   slicfunc.cpp@704 : Slic_EyePoint: (45,100,0)
    objpool.cpp@92  : ObjPool: Inserting object id 0xc0000004
    c3debug.cpp@303 : Assertion (m_traderUnits->Num() > 0) Failed in File:c:\activision\civctp2\ctp2_code\gs\gameobj\player.cpp, Line:3246
    c3debug.cpp@304 : Stack Trace: '  0x0149f5ec  [char * __cdecl c3debug_StackTrace(void) + 0x3c]
  0x01497342  [void __cdecl c3debug_Assert(char const *,char const *,int) + 0x72]
  0x00cf597f  [public: void __thiscall Player::KillATrader(void) + 0x6f]
  0x00cf9761  [public: void __thiscall Player::RemoveTradeRoute(class TradeRoute,enum CAUSE_KILL_TRADE_ROUTE) + 0x91]
  0x00d23f0c  [public: void __thiscall TradeRoute::KillRoute(enum CAUSE_KILL_TRADE_ROUTE) + 0x3bc]
  0x00c7981f  [public: void __thiscall TradeRoute::Kill(enum CAUSE_KILL_TRADE_ROUTE) + 0x2f]
  0x00d1f6d9  [private: virtual enum GAME_EVENT_HOOK_DISPOSITION __thiscall KillTradeRouteEvent::GEVHookCallback(enum GAME_EVENT,class GameEventArgList *) + 0x69]
  0x00bdb3fa  [public: enum GAME_EVENT_ERR __thiscall GameEventHook::Activate(class GameEventArgList *,int,int &)const  + 0xfa]
  0x00bde856  [public: enum GAME_EVENT_ERR __thiscall GameEventManager::ActivateHook(enum GAME_EVENT,class GameEventArgList *,int,int &) + 0x86]
  0x00bd6175  [public: enum GAME_EVENT_ERR __thiscall GameEvent::Process(void) + 0x65]
  0x00be1926  [public: enum GAME_EVENT_ERR __thiscall GameEventManager::ProcessHead(void) + 0x76]
  0x00be183e  [public: enum GAME_EVENT_ERR __thiscall GameEventManager::Process(void) + 0x8e]
  0x014765b1  [public: int __thiscall CivApp::InitializeGame(class CivArchive &) + 0x6d1]
  0x00bf2cb3  [public: unsigned int __thiscall GameFile::Restore(char const *) + 0xd63]
  0x00bf3854  [public: static void __cdecl GameFile::RestoreGame(char const *) + 0x34]
  0x01478c09  [public: int __thiscall CivApp::LoadSavedGame(char const *) + 0x159]
  0x01471b75  [public: virtual void __thiscall LoadSaveGameAction::Execute(class aui_Control *,unsigned int,unsigned int) + 0x35]
  0x01175461  [public: void __thiscall aui_UI::HandleActions(void) + 0x81]
  0x01178636  [public: virtual enum AUI_ERRCODE __thiscall aui_UI::Process(void) + 0x86]
  0x0147a84a  [private: int __thiscall CivApp::ProcessUI(unsigned int,unsigned int &) + 0x27a]
  0x01479d0f  [public: int __thiscall CivApp::Process(void) + 0x10f]
  0x01466b0b  [int __stdcall CivMain(struct HINSTANCE__ *,struct HINSTANCE__ *,char *,int) + 0x41b]
  0x0146b97e  [WinMain@16 + 0xae]
  0x014ba97e  [int __cdecl invoke_main(void) + 0x1e]
  0x014ba817  [int __cdecl __scrt_common_main_seh(void) + 0x157]
  0x014ba6ad  [int __cdecl __scrt_common_main(void) + 0xd]
  0x014ba9f8  [WinMainCRTStartup + 0x8]
  0x74e06359  [$R002030 + 0x73622fc9]
  0x775a7b74  [$R002030 + 0x75dc47e4]
'
```

One object is deleted before it actually had been added to the pool. This doesn't seem to be right. And the assert seems that the trade route is killed before it was fully created, so that it does not have any traders assigned, yet.

Keep in mind that could be a problem that could be present in Activision's version. Just nobody dared to load that scenario with the debug version.

**Comment by LynxAbraxas** [comment]
Did not have the time to dig into this, but it seems the catamaran does work in the samurai scenario, but there it can transport a unit.

**Comment by MartinGuehmann** [comment]
> Did not have the time to dig into this, but it seems the catamaran does work in the samurai scenario, but there it can transport a unit.

Are we talking about the last commit? Actually, that is about the Alex scenario, where they turned most transporters into battleships.

File: Scenarios/AlexanderTheGreat/scen0000/default/aidata/UnitBuildLists.txt [hunk 1]
```diff
@@ -94,7 +94,7 @@ UNIT_BUILD_LIST_SPECIAL_NUCLEAR {
 ## 12 #########################################################
 
 UNIT_BUILD_LIST_SEA_TRANSPORT {
-  Unit UNIT_CATAMARAN
+  Unit UNIT_TRIREME
```
Comment (LynxAbraxas) [inline Scenarios/AlexanderTheGreat/scen0000/default/aidata/UnitBuildLists.txt:97]: > Actually, that is about the Alex scenario, where they turned most transporters into battleships.

I see. Yes it was about this change.

**Comment by LynxAbraxas** [comment]
> One object is deleted before it actually had been added to the pool. This doesn't seem to be right. And the assert seems that the trade route is killed before it was fully created, so that it does not have any traders assigned, yet.

Yes, I had this problem before, fixed it for the general game play but the scenarios might need some special handling, but I don't get there. For me loading alex ends (apparently not segfaulting) after many
```
Assertion (start < m_code + m_codeSize) Failed in File:SlicSegment.cpp, Line:649
```
(which can be ignored) with
```
Illegal instruction
```
which seems very much like what @madand got as well.

**Comment by MartinGuehmann** [comment]
> > One object is deleted before it actually had been added to the pool. This doesn't seem to be right. And the assert seems that the trade route is killed before it was fully created, so that it does not have any traders assigned, yet.
> 
> Yes, I had this problem before, fixed it for the general game play but the scenarios might need some special handling, but I don't get there. For me loading alex ends (apparently not segfaulting) after many
> 
> ```
> Assertion (start < m_code + m_codeSize) Failed in File:SlicSegment.cpp, Line:649
> ```

Well, that is a Linux thing, maybe a 64 bit thing. On Linux the debug version gives me this stack trace:

```
    c3debug.cpp@303 : Assertion (start < m_code + m_codeSize) Failed in File:SlicSegment.cpp, Line:649
    c3debug.cpp@304 : Stack Trace: '  0x6c9da5e4  [DebugMemory_GuardedBlockRealloc(char const*, int, MemoryHeapDescriptor*, void*, unsigned long) + 0x19f]
  0x6c9da90e  [Debug_MemNodeList(int) + 0x12]
  0x6c9d796d  [(anonymous namespace)::c3files_GetVolumeName(int) + 0x10d]
  0x6cf05d11  [gameinit_Initialize(int, int, CivArchive&) + 0x1866]
  0x6cf05df9  [gameinit_Initialize(int, int, CivArchive&) + 0x194e]
  0x6cf05b52  [gameinit_Initialize(int, int, CivArchive&) + 0x16a7]
  0x6ccf28e5  [Army::IsEnemy(Unit) const + 0x45]
  0x6ccf0f9c  [AgreementData::RecipientIsViolating(int, unsigned int) + 0x56]
  0x6cf056ef  [gameinit_Initialize(int, int, CivArchive&) + 0x1244]
  0x6cf04fa0  [gameinit_Initialize(int, int, CivArchive&) + 0xaf5]
  0x6cebd681  [SlicFrame::SlicFrame(SlicSegment*, int) + 0x47]
  0x6cea9f50  [SlicDBConduit<WonderMovieRecord, WonderMovieRecordAccessorInfo>::GetValue(int, char const*, int) + 0xa6]
  0x6cea9a85  [SlicDBConduit<CivilisationRecord, CivilisationRecordAccessorInfo>::GetRecordNameIDByIndex(int) + 0x6b]
  0x6cf24231  [World::CalcCumScore(int, int, int, float&, float**) + 0x6b]
  0x6c9e6957  [CivApp::InitializeSpriteEditor(CivArchive&) + 0x737]
  0x6ccf7b70  [ArmyData::InciteRevolution(MapPoint const&) + 0xd2]
  0x6ccf5f98  [ArmyData::CityRadiusFunc(MapPoint const&) + 0x178]
  0x6c9e8de6  [void allocated::clear<SaveInfo>(SaveInfo*&) + 0x4]
  0x6c9e9926  [get_name(char const*) + 0x24f]
  0x6ca6cc84  [aui_Window::MouseLGrabInside(aui_MouseEvent*) + 0x64]
  0x6ca6cd72  [aui_Window::DeleteSurfaceIfDynamic() + 0x32]
  0x6c9e8674  [void allocated::clear<OzoneDatabase>(OzoneDatabase*&) + 0x32]
  0x6c9e8b14  [void allocated::clear<CTPDatabase<EndGameObjectRecord> >(CTPDatabase<EndGameObjectRecord>*&) + 0xe]
  0x6c9e099f  [aui_Resource<Picture>::AddSearchPath(char const*) + 0x53]
  0x6c9e05e2  [aui_Resource<aui_BitmapFont>::AddSearchPath(char const*) + 0xda]
  0x693b81e3  [(unknown) + 0x0]
  0x6c9cb85e  [ArmyClumps::Execute(int, char**) + 0x10]
'
```
We are at the place where it fails but the stack trace is just nonsense. Looks like that it reads an outdated ctp2linux.map. I can find an old one but not the one that is supposed to be created with the executable. The map file is supposed to be in the executable folder: ctp2_code/ctp. By the way is there also a way that the executable ctp2 is build in the executable folder and not one above. That is not very nice, I can set in KDevelop the right folder as working dir, but still this is not nice.

**Comment by LynxAbraxas** [comment]
>  Looks like that it reads an outdated ctp2linux.map. I can find an old one but not the one that is supposed to be created with the executable.

`ctp2linux.map` creation was changed in 1c4073e07c708c7326721866be14d3225df6df17 as part of #107. So it should appear with a debug-build. It is part of the last debug-DF under `/opt/ctp2/ctp2_program/ctp`:
```
docker run registry.gitlab.com/civctp2/civctp2/master:test-debug ls -la ctp2linux.map
-rw-r--r-- 1 root root 5071674 Mar 24 21:28 ctp2linux.map
```

> By the way is there also a way that the executable ctp2 is build in the executable folder and not one above.

Never bothered and therefore never tried to change that, should be possible though.

**Comment by MartinGuehmann** [comment]
> > Looks like that it reads an outdated ctp2linux.map. I can find an old one but not the one that is supposed to be created with the executable.
> 
> `ctp2linux.map` creation was changed in [1c4073e](https://github.com/civctp2/civctp2/commit/1c4073e07c708c7326721866be14d3225df6df17) as part of #107. So it should appear with a debug-build. It is part of the last debug-DF under `/opt/ctp2/ctp2_program/ctp`:
> 
> ```
> docker run registry.gitlab.com/civctp2/civctp2/master:test-debug ls -la ctp2linux.map
> -rw-r--r-- 1 root root 5071674 Mar 24 21:28 ctp2linux.map
> ```

Well, it should be in /ctp2_code/ctp. And as far as I can see it is created when you "install" it. Of course that is something you don't do when you compile it from an IDE. And of course why should you? All the files are in place, so you just have to start the executable.

> What is indicating that it is reading an outdated ctp2linux.map?

That's simple, I have a ctp2linux.map file in /ctp2_code/ctp/. However, it is from Feb 2019, but I would expect one that came from the executable I have just built.

> 
> > We are at the place where it fails but the stack trace is just nonsense.
> 
> Not sure, why should it be nonsense? It must come from some state of the game.

You have the function addresses that come from the executable and you have the names of the functions that come from the ctp2linux.map file. Since, it is outdated the assignments are off. For instance SlicFrame::SlicFrame does not call gameinit_Initialize. And if there was no ctp2linux.map these things would show up as unknown.

In principle the stack trace could be implemented with some Linux specific stack trace function, but that wouldn't work for debug memory.

**Comment by LynxAbraxas** [comment]
I see what you mean now. Never had that problem so far because I basically use a commit like 4bb80b1cd297a5ce0312181bc10f32d940e17964 to debug directly inside the docker container with gdb. 
Which IDE are you using under linux? Perhaps you can instruct it to run
https://github.com/civctp2/civctp2/blob/f8f96b23b0ab7c30ae248bec04bb3ad33f04f6cc/GNUmakefile.am#L17
after compilation?

**Comment by MartinGuehmann** [comment]
> Which IDE are you using under linux?

It is KDevelop. It just uses make for making everything. So you have to fix it in the makefiles. Which actually #304 does.

Do you also have some tips about gdb, which is the debugger KDevelop uses. I can only use it with c source files but not with c++.

**Comment by LynxAbraxas** [comment]
> Do you also have some tips about gdb, which is the debugger KDevelop uses. I can only use it with c source files but not with c++.

My knowledge about `gdb` is limited, but it did always work for me even for c++ files. However, I always used it from the command line, with commands like in 2564ca27c0a1dc84cc31d038f5b1fed92859f030 run like b1d375a299ef17189bd67ba7f6ab7aaacfdf5516 if compiled with `-ggdb` for C and CXX as in 557f4a0d148673e1d6cd0a72782fd3e671938245.

**Comment by LynxAbraxas** [comment]
> I can only use it with c source files but not with c++.

What is happening (i.e. not working) for you when using it with c++?

**Comment by MartinGuehmann** [comment]
> > I can only use it with c source files but not with c++.
> 
> What is happening (i.e. not working) for you when using it with c++?

When I click on the function in the stack trace I get to the source file if it is a c-file, but not if it is a cpp-file.

**Comment by LynxAbraxas** [comment]
> When I click on the function in the stack trace I get to the source file if it is a c-file, but not if it is a cpp-file.

Hm, that is very specific. Have you set `-ggdb` for C++ e.g. in `CXXFLAGS`?
Does it work if you run gdb from command line like `gdb -ex run --args ./ctp2 nointromovie`?

**Comment by LynxAbraxas** [comment]
With #312 the only scenario that exhibits problems for me is the AE scenario with religion mods, yielding e.g.:
```
Messagebox(DB Error): UNIT_BUILD_LIST_SPECIAL_SLAVERY not found in UnitBuildList database

Continue?
Messagebox(DB Error): UNIT_BUILD_LIST_SPECIAL_DIPLOMACY not found in UnitBuildList database

Continue?
Messagebox(DB Error): UNIT_BUILD_LIST_SPECIAL_MISSIONARY not found in UnitBuildList database

Continue?
Messagebox(DB Error): UNIT_BUILD_LIST_SPECIAL_SLAVERY not found in UnitBuildList database

Continue?
Messagebox(DB Error): UNIT_BUILD_LIST_SPECIAL_DIPLOMACY not found in UnitBuildList database
```
otherwise this I guess is read for merge.

**Comment by LynxAbraxas** [comment]
NB: There is nothing listed for me in the "Weltkarten" scenario.

**Comment by MartinGuehmann** [comment]
Abandoned in favor of #312.

**Pull request closed without merging**